src/BenchmarkDotNet/Configs/ConfigExtensions.cs

```diff
@@ -18,38 +18,90 @@ namespace BenchmarkDotNet.Configs
         /// <summary>
         /// determines if the exported result files should not be overwritten (be default they are overwritten)
         /// </summary>
-        [PublicAPI] public static IConfig DontOverwriteResults(this IConfig config, bool value = true) => config.With(m => m.Options = m.Options.Set(value, ConfigOptions.DontOverwriteResults));
+        [Obsolete("This property will soon be removed, please start using WithOptionsIf instead.")]
+        [PublicAPI] public static IConfig DontOverwriteResults(this IConfig config, bool value = true) => WithOptionsIf(config, value, ConfigOptions.DontOverwriteResults);
         /// <summary>
         /// determines if benchmarking should be stopped after the first error (by default it's not)
         /// </summary>
-        [PublicAPI] public static IConfig StopOnFirstError(this IConfig config, bool value = true) => config.With(m => m.Options = m.Options.Set(value, ConfigOptions.StopOnFirstError));
-        [PublicAPI] public static IConfig WithArtifactsPath(this IConfig config, string artifactsPath) => config.With(m => m.ArtifactsPath = artifactsPath);
+        [Obsolete("This property will soon be removed, please start using WithOptionsIf instead.")]
+        [PublicAPI] public static IConfig StopOnFirstError(this IConfig config, bool value = true) => WithOptionsIf(config, value, ConfigOptions.StopOnFirstError);
+
         /// <summary>
         /// sets given options for the config
         /// </summary>
-        [PublicAPI] public static IConfig With(this IConfig config, ConfigOptions options) => config.With(m => m.Options = options);
-        [PublicAPI] public static IConfig With(this IConfig config, params BenchmarkLogicalGroupRule[] rules) => config.With(c => c.Add(rules));
+        [Obsolete("This property will soon be removed, please start using WithOptions instead.")]
+        [PublicAPI] public static IConfig With(this IConfig config, ConfigOptions options) => WithOptions(config, options);
+        [PublicAPI] public static ManualConfig WithOptions(this IConfig config, ConfigOptions options) => config.With(m => m.WithOptions(options));
+        [PublicAPI] public static ManualConfig WithoutOptions(this IConfig config, ConfigOptions options) => config.With(m => m.WithoutOptions(options));
+        [PublicAPI] public static ManualConfig WithOptionsIf(this IConfig config, bool value, ConfigOptions options) => config.With(m => m.WithOptionsIf(value, options));
+        
+
+        [Obsolete("This property will soon be removed, please start using GroupBenchmarksBy instead.")]
+        [PublicAPI] public static IConfig With(this IConfig config, params BenchmarkLogicalGroupRule[] rules) => GroupBenchmarksBy(config, rules);
+        [PublicAPI] public static ManualConfig GroupBenchmarksBy(this IConfig config, params BenchmarkLogicalGroupRule[] rules) => config.With(c => c.GroupBenchmarksBy(rules));
```

Comment:
I think we should use the `Add*` notation here. E.g., it can be renamed to `AddLogicalGroupRule`.

Reply:
@AndreyAkinshin I've renamed it to `WithLogicalGroupRules`.